src/parsy/__init__.py

```diff
@@ -12,13 +12,70 @@
 noop = lambda x: x
 
 
+class StrStream(str):
+    """String data to parse, possibly equipped with a name for the source it's
+    from, e.g. a file path."""
+
+    def __new__(cls, string, source):
+        instance = super().__new__(cls, string)
+        instance.source = source
+        return instance
+
+
+class ByteStream(str):
+    """String data to parse, possibly equipped with a name for the source it's
+    from, e.g. a file path."""
+
+    def __new__(cls, bs, source):
+        instance = super().__new__(cls, bs)
+        instance.source = source
+        return instance
+
+
+def make_stream(data: str | bytes, source: Any):
+    """Constructs an appropriate stream type for `data` when it's one of the
+    three core supported datatypes of parsy (viz. str, bytes, list). Otherwise,
+    the data is assumed to just support a minimum of __getitem__ and
+    __len__."""
+    if isinstance(data, str):
+        return StrStream(data, source)
+
+    if isinstance(data, bytes):
+        return ByteStream(data, source)
+
+    raise RuntimeError(
+        "A Parsy stream can be formed only on str and bytes, but the given "
+        f"data has type {type(data)}. If you are separately tokenizing the "
+        "data to parse, consider instead equipping the tokens with source "
+        "location metadata.",
+    )
```

Comment:
This doesn't look like it handles `list`, in contrast to its docstring. That's OK, but the docstring should be corrected, and preferably the limitation should be noted in the appropriate docs i.e. the parse method.